### PR TITLE
New algorithm to make ins file for shelx - SaveINS

### LINKED ISSUE #34275: Produce a .ins (instrument) input file for SHELX
**Is your feature request related to a problem? Please describe.**
SHELX requires a .ins file as input fro single-crystal refinements (possibly also powder) - currently mantid does not produce this
Reported by Nick Funnell

**Describe the solution you'd like**
Mantid would produce a .ins file

**Describe alternatives you've considered**
Currently hand-made by each user - quite a lengthy process.

**Additional context**
The specification for a .ins file are as follows from top line to bottom of attached file  [roy40_3d.txt](https://github.com/mantidproject/mantid/files/9290524/roy40_3d.txt):
-	A title line
-	A comment line (denoted by ‘REM’)
-	CELL - Cell dimensions with a dummy wavelength value (0.7) for whitebeam
-	ZERR - Number of formula units in the cell ‘Z’ (=4) and the errors ‘ERR’ on the cell dimensions
-	LATT instruction – a number code for the centring. 1 = P, 2 = I, 3 = R, etc. A negative value indicates non-centrosymmetric
-	SYMM – symmetry operators for space group number 14 P21/n. You’ll see more operations defined in the international tables (4 of them) but one line is all you need here; ‘x, y, z’ is assumed. Then if you combine these two operations with the inversion symmetry indicated by the LATT instruction, then you can generate all 4 SYMM positions. 
-	NEUT – this is neutron data
-	SFAC – this specifies the scattering factors for each atom type, with values corresponding to neutrons. I’m not 100% sure if this detail is still required because I think the NEUT instruction should deal with this. It could be a hangover of earlier versions of the program. Silvia will know better than me.
-	UNIT – number of each atom type in the unit cell 
-	MERG 0 – instruction not to merge the data (otherwise default is to merge symmetry-equivalent reflections)
-	HKLF2 - .hkl file format 2
-	END 

Documentation: https://cpb-us-w2.wpmucdn.com/web.sas.upenn.edu/dist/f/266/files/2016/11/shelxl_comlist-1mp04in.pdf

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

New algorithm to make ins file for shelx - see algorithm doc file for details.

**To test:**

(1) Run this script
```
from mantid.simpleapi import *

LoadEmptyInstrument(InstrumentName='SXD', OutputWorkspace='SXD')
SetUB(Workspace='SXD', a=7.6508, b=13.2431, c=11.6243, alpha=90, beta=104.1183, gamma=90)
ndensity = 4/mtd['SXD'].sample().getOrientedLattice().volume()
SetSample(InputWorkspace='SXD', Material={'ChemicalFormula': 'C12 H9 N3 O2 S1', 'SampleNumberDensity': ndensity})
SaveINS(InputWorkspace='SXD', Filename=r"C:\Users\xhg73778\Desktop\test.ins", Spacegroup='P 1 21/n 1')
SaveINS(InputWorkspace='SXD', Filename=r"C:\Users\xhg73778\Desktop\test_isotope.ins", Spacegroup='P 1 21/n 1', UseNaturalIsotopicAbundances=False)
```
(2) Check the files are saved
- The `test,ins` file should have a single `SFAC` line (`SFAC C H N O S`)
- The `test_isotope.ins` should print out more info (scattering lengths etc.) for each atom.

Fixes #34275 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/SaveINS-v1.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+The ``InputWorkspace`` must have a UB set (from which the lattice parameters are retrieved) and a sample material
+(from which the atoms and number of formula units are taken). The spacegroup can be taken given as an input argument to
+the function, or from the crystal structure if one has been set on the workspace.
+
+By default it is assumed that atoms have a ratio of isotopes seen in nature, however if
+``UseNaturalIsotopicAbundances=False`` the scattering scattering cross-sections defined for the isotopes specified in
+the sample material will be used. In this case there are other atomic properties that are output, one of these is the
+covalent radius (which SHELX uses for the purposes of outputting bonds/geometry information) which is set to be
+1 Angstrom (if ``UseNaturalIsotopicAbundances=True`` SHELX will look up the covalent radii of atoms itself).
```
Comment (jhaigh0) [inline docs/source/algorithms/SaveINS-v1.rst:24]: Is there a test for this case where `covalent radius` is different depending on `UseNaturalIsotopicAbundances`?

File: Framework/PythonInterface/plugins/algorithms/SaveINS.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            if not SpaceGroupFactory.isSubscribedSymbol(spgr_sym):
+                issues["Spacegroup"] = "Not a valid spacegroup symbol."
+        elif not ws.sample().hasCrystalStructure:
+            issues["InputWorkspace"] = "The workspace does not have a crystal structure defined, a spacegroup must " "be provided as input."
+        # check the workspace has a UB defined
+        if not sample.hasOrientedLattice():
+            issues["InputWorkspace"] = "Workspace must have an oriented lattice defined."
+        if sample.getMaterial().numberDensity < 1e-15:
+            issues["InputWorkspace"] = "Workspace must have a sample material set."
+        return issues
```
Comment (jhaigh0) [inline Framework/PythonInterface/plugins/algorithms/SaveINS.py:74]: Does this need to validate the `Filename` property to check the file exists?

**Review by jhaigh0 — CHANGES_REQUESTED** [review]
Works exactly as the tests expect and the code is very readable 👍 Left a few suggests / comments for clarification, but nothing major

File: Framework/PythonInterface/test/python/plugins/algorithms/SaveINSTest.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        SetSample(InputWorkspace=self.ws, Material={"ChemicalFormula": "C12 H9 N3 O2 S1", "SampleNumberDensity": ndensity})
+
+    def tearDown(self):
+        shutil.rmtree(self._tmp_directory)
+
+    def test_save_ins_throws_if_no_oriented_lattice(self):
+        ClearUB(self.ws)
+        output_file = path.join(self._tmp_directory, "test.ins")
+
+        self.assertRaises(RuntimeError, SaveINS, Filename=output_file, InputWorkspace=self.ws, Spacegroup="P 1 21/n 1")
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/test/python/plugins/algorithms/SaveINSTest.py:52]: Thanks that's a great suggestion, I didn't know how to do that!

File: docs/source/algorithms/SaveINS-v1.rst (hunk 1 above)
Comment (RichardWaiteSTFC) [reply docs/source/algorithms/SaveINS-v1.rst:24]: So the covalent radius is only ever printed to file when `UseNaturalIsotopicAbundances=False` (one of the numbers in the `SFAC` line) - so this is covered by `test_save_ins_natural_isotopic_abundance_true`

File: Framework/PythonInterface/test/python/plugins/algorithms/SaveINSTest.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhaigh0) [reply Framework/PythonInterface/test/python/plugins/algorithms/SaveINSTest.py:52]: Found it [assertRaisesRegex](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex)

File: docs/source/algorithms/SaveINS-v1.rst (hunk 1 above)
Comment (jhaigh0) [reply docs/source/algorithms/SaveINS-v1.rst:24]: That makes sense thanks

**Review by jhaigh0 — APPROVED** [review]

**Pull request merged**